io-sim-classes/src/Control/Monad/Class/MonadSTM/Strict.hs

```diff
@@ -76,11 +76,17 @@ type LazyTMVar m = Lazy.TMVar m
   Strict TVar
 -------------------------------------------------------------------------------}
 
+#if CHECK_TVAR_INVARIANT
 data StrictTVar m a = StrictTVar
    { invariant :: !(a -> Maybe String)
      -- ^ Invariant checked whenever updating the 'StrictTVar'.
    , tvar      :: !(LazyTVar m a)
    }
+#else
```

Comment:
Why the emphasizes on using a newtype instead of data in this particular case?

Reply:
`newtype`s do not change runtime representation; By moving from `data` with two fields to a `newtype` we save 3 machine words and one indirection (pointer) to the actual data that is interesting: the `tvar`.

Reply:
ah TIL! But is this saving worth the complexity? Should this be done whenever possible?

Reply:
Actually the complexity is quite low because the constructor is not exported making the impact isolated to a single module. Building both versions of `io-sim-classes` on `ci` shouldn't be too difficult too.

Reply:
Yeah, that makes sense! Nice, thanks :D

Reply:
Without knowing any details, I would guess this won't make an observable difference in any performance metrics. So Armando's question about complexity is the first that came to my mind, too. But I was worried about "the less CPP the better" more so than encapsulation concerns etc.

I'm not objecting, just sharing my opinion. The extra CPP here is not overwhelming.

We could get by entirely without CPP by using a conditional in the cabal file to choose between two very short source files that both define `StrictTVar` and its fundamental interface. That appeals to me, but also seems a bit extreme in the other direction.

Reply:
The performance impact will not be visible, we'd need to have 100s if not 1000s of `StrictTVar` for it to be visible in some way; but I think if we want to put it on hackage we should make the change.

> We could get by entirely without CPP by using a conditional in the cabal file to choose between two very short source files that both define StrictTVar and its fundamental interface. That appeals to me, but also seems a bit extreme in the other direction.

Yes that's another option, though it would not make a difference: one needs to compile the package twice to type check all the code.  We could go around this with two modules and a CPP declaration used to import one of the representations, but it feels like we'll trying to be to clever for something that's quite simple.